bfcfg

```diff
@@ -188,6 +202,353 @@ part_info()
   done
 }
 
+#
+# Add header into a boot entry
+# $1: output file
+#
+boot_cfg_add_header()
+{
+  printf "\\x07\\x00\\x00\\x00\\x01\\x00\\x00\\x00" >> $1
+}
+
+#
+# Add length into a boot entry
+# $1: output file
+# $2: length
+#
+boot_cfg_add_len()
+{
+  len=$2
+  len=$(printf '%04x' ${len})
+  printf "\\x${len:2:2}\\x${len:0:2}" >> $1
+}
+
+#
+# Add tail into a boot entry
+# $1: output file
+#
+boot_cfg_add_tail()
+{
+  printf "\\x7f\\xff\\x04\\x00" >> $1
+}
+
+#
+# Add name into a boot entry
+# $1: output file
+# $2: name
+#
+boot_cfg_add_name()
+{
+  local idx name=$2
+
+  for ((idx=0; idx<${#name}; idx++)); do
+    printf "${name:$idx:1}" >> $1
+    printf "\\x00" >> $1
+  done
+  printf "\\x00\\x00" >> $1
+}
+
+#
+# Add PCIe info into a boot entry
+# $1: output file
+# $2: port name
+#
+boot_cfg_add_pcie()
+{
+  printf \
+"\\x02\\x01\\x0c\\x00\\xd0\\x41\\x03\\x0a\\x00\\x00\\x00\\x00"\
+"\\x01\\x01\\x06\\x00\\x00\\x00\\x01\\x01\\x06\\x00\\x00\\x00"\
+"\\x01\\x01\\x06\\x00\\x00\\x02\\x01\\x01\\x06\\x00" >> $1
+  if [ "$2" = "NIC_P0" ]; then
+    printf "\\x00" >> $1
+  else
+    printf "\\x01" >> $1
+  fi
+  printf "\\x00" >> $1
+}
+
+#
+# Add Eth/MAC info into a boot entry
+# $1: output file
+# $2: MAC address (xx:xx:xx:xx:xx:xx)
+#
+boot_cfg_add_eth()
+{
+  local idx mac=$2
+
+  printf "\\x03\\x0b\\x25\\x00" >> $1
+  mac="\\x$(echo ${mac//:/\\x})"
+  printf ${mac} >> $1
+  for idx in {1..26}; do
+    printf "\\x00" >> $1
+  done
+  printf "\\x01" >> $1
+}
+
+#
+# Add VLAN info into a boot entry
+# $1: output file
+# $2: decimal VLAN id
+#
+boot_cfg_add_vlan()
+{
+  local vlan_id=$2
+
+  vlan_id=$(printf '%04x' ${vlan_id})
+  printf "\\x03\\x14\\x06\\x00\\x${vlan_id:2:2}\\x${vlan_id:0:2}" >> $1
+}
+
+#
+# Add IPv4 info into a boot entry
+# $1: output file
+#
+boot_cfg_add_ipv4()
+{
+  local idx
+
+  printf "\\x03\\x0c\\x1b\\x00" >> $1
+  for idx in {1..23}; do
+    printf "\\x00" >> $1
+  done
+}
+
+#
+# Add IPv6 info into a boot entry
+# $1: output file
+#
+boot_cfg_add_ipv6()
+{
+  local idx
+
+  printf "\\x03\\x0d\\x3c\\x00" >> $1
+  for idx in {1..39}; do
+    printf "\\x00" >> $1
+  done
+  printf "\\x40" >> $1
+  for idx in {1..16}; do
+    printf "\\x00" >> $1
+  done
+}
+
+#
+# Create boot entry for UEFI shell
+# $1: output file
+# $2: name
+#
+boot_cfg_create_shell()
+{
+  boot_cfg_add_header $1
+  boot_cfg_add_len $1 48
+  boot_cfg_add_name $1 "EFI Internal Shell"
+  printf \
+"\\x01\\x03\\x18\\x00\\x0b\\x00\\x00\\x00\\x00\\xf0\\x4f\\xfe\\x00\\x00\\x00\\x00"\
+"\\x7f\\x62\\xa6\\xfe\\x00\\x00\\x00\\x00\\x04\\x06\\x14\\x00\\x83\\xa5\\x04\\x7c"\
+"\\x3e\\x9e\\x1c\\x4f\\xad\\x65\\xe0\\x52\\x68\\xd0\\xb4\\xd1" >> $1
+  boot_cfg_add_tail $1
+}
```

Comment:
Note sure if this is valid across bootloaders, the shell device path may change

Reply:
Good catch! Another edk2 code review request has been sent to address this issue. It creates or fix the UEFI shell entries. (https://bu-gerrit2.mtbu.labs.mlnx/#/c/edk2/+/6127/)